### ISSUE TITLE
mailer does not wait for HELO (does not conform to RFCs)

### ISSUE THREAD

**Issue by ghost** [Marak/node_mailer#12]
Hi,

Just an FYI, I've noticed that mailer does not conform to RFCs by not waiting for the HELO before proceeding to send data to the SMTP server.

I am getting the following error when I try to send to an exim4 server.  This problem is indicative of not waiting for HELO before sending.

```
SMTP protocol synchronization error (input sent without waiting for greeting)
```

I would like to use mailer but I will need to look for a different solution until this problem is fixed.

Cheers,

Matt

**Comment by Marak** [comment]
Matt -

I think we could make that modification fairly easily....

The biggest issue I've had with this library is testing. Do you have any ideas on how we could add test coverage for this issue? I've been using sendmail and pythons built in smtp server and haven't seen that issue come up. I'm not 100% familiar with the smtp RFC, but I do believe you.
- Marak

**Comment by mikegerwitz** [comment]
I get this error while attempting to send mail via exim4. Here's an example of the issue (see the answer):

http://stackoverflow.com/questions/2264673/554-smtp-synchronization-error-with-exim4-and-my-code

**Comment by Marak** [comment]
Yes, this needs to be fixed in node_mailer v0.5.0.

closing this ticket and merging with https://github.com/Marak/node_mailer/issues#issue/17

**Comment by radicaldynamic** [comment]
Marak - 

SMTP is a beast to test due to the vast number of implementations  I am not aware of any test suites off the top of my head but following the RFC should give a fairly good start.  The trouble is at that point you're almost _building_ an SMTP server.

I think I have switched to a different module at this point but you might want to add an option to send the mail to /usr/sbin/sendmail (or other) if you haven't already.  That will provide another way for people to send in the event that their SMTP server cannot be reached (or is just plain not working with node_mailer).

Just my two cents,

Matt

**Comment by Marak** [comment]
this actually should have been fixed in 0.4.5, just forgot to close ticket. cheers!
